File: know_engine_py/app/schemas/document.py
from pydantic import BaseModel, ConfigDict, Field

from know_engine_py.app.rag.splitters.types import DocumentSplitParam, SplitType


class DocumentResponse(BaseModel):
    """文档接口响应模型，对应 KnowledgeDocument 的前端可见字段。"""

    model_config = ConfigDict(from_attributes=True)

    doc_id: int
    doc_title: str
    upload_user: str | None = None
    doc_url: str | None = None
    converted_doc_url: str | None = None
    status: str
    accessible_by: str | None = None
    description: str | None = None
    knowledge_base_type: str | None = None
    group_id: str | None = None
    knowledge_base_id: str | None = None
    file_object_id: str | None = None
    extension: dict | None = None


class SegmentResponse(BaseModel):
    """文档分段响应模型，对应 KnowledgeSegment 的前端可见字段。"""

    model_config = ConfigDict(from_attributes=True, populate_by_name=True)

    segment_id: int = Field(alias="id")
    document_id: int
    chunk_id: str | None = None
    chunk_order: int
    text: str
    status: str
    skip_embedding: int
    extra_metadata: dict | None = None


class DocumentSplitRequest(BaseModel):
    """触发文档切分的请求体。"""

    split_type: SplitType = Field(default=SplitType.TITLE, alias="splitType")
    chunk_size: int = Field(default=800, alias="chunkSize", gt=0)
    overlap: int = Field(default=80, ge=0)
    title_level: int = Field(default=1, alias="titleLevel", gt=0)
    regex: str | None = None
    separator: str | None = None

    model_config = ConfigDict(populate_by_name=True)

    def to_split_param(self) -> DocumentSplitParam:
        """转换成 splitter 层内部参数对象，避免 service 依赖 API schema。"""
        return DocumentSplitParam(
            split_type=self.split_type,
            chunk_size=self.chunk_size,
            overlap=self.overlap,
            title_level=self.title_level,
            regex=self.regex,
            separator=self.separator,
        )


class DocumentSplitResponse(BaseModel):
    """文档切分响应：包含切分数量和后续索引任务投递结果。"""

    model_config = ConfigDict(populate_by_name=True)

    document_id: int = Field(alias="documentId")
    segment_count: int = Field(alias="segmentCount")
    index_queued: bool = Field(alias="indexQueued")
    index_task_id: str | None = Field(default=None, alias="indexTaskId")
    index_queue_error: str | None = Field(default=None, alias="indexQueueError")


class DocumentImportResponse(DocumentResponse):
    """文档导入响应：包含导入结果和后续转换任务投递结果。"""

    model_config = ConfigDict(from_attributes=True, populate_by_name=True)

    conversion_queued: bool = Field(default=False, alias="conversionQueued")
    conversion_task_id: str | None = Field(default=None, alias="conversionTaskId")
    conversion_queue_error: str | None = Field(default=None, alias="conversionQueueError")

File: know_engine_web/src/api/document.ts
export interface KnowledgeDocument {
  doc_id: number;
  doc_title: string;
  upload_user?: string | null;
  doc_url?: string | null;
  converted_doc_url?: string | null;
  status: string;
  accessible_by?: string | null;
  description?: string | null;
  knowledge_base_type?: string | null;
  group_id?: string | null;
  knowledge_base_id?: string | null;
  file_object_id?: string | null;
  extension?: Record<string, unknown> | null;
  conversionQueued?: boolean;
  conversionTaskId?: string | null;
  conversionQueueError?: string | null;
}

export interface DocumentSplitResult {
  documentId: number;
  segmentCount: number;
  indexQueued: boolean;
  indexTaskId?: string | null;
  indexQueueError?: string | null;
}

export interface KnowledgeSegment {
  id?: number;
  segment_id?: number;
  document_id: number;
  chunk_id?: string | null;
  chunk_order: number;
  text: string;
  status: string;
  skip_embedding: number;
  extra_metadata?: Record<string, unknown> | null;
}

export interface DocumentTask {
  taskId: string;
  documentId: number;
  taskType: string;
  status: string;
  celeryTaskId?: string | null;
  currentAttempt: number;
  maxAttempts: number;
  lastError?: string | null;
  startedAt?: string | null;
  finishedAt?: string | null;
  nextRetryAt?: string | null;
  taskMetadata?: Record<string, unknown> | null;
  createdAt: string;
  updatedAt: string;
}

export interface ImportDocumentRequest {
  baseUrl: string;
  accessToken: string;
  knowledgeBaseId: string;
  file: File;
  docTitle?: string;
  description?: string;
}

export interface ListDocumentsRequest {
  baseUrl: string;
  accessToken: string;
  knowledgeBaseId: string;
}

export interface SplitDocumentRequest {
  baseUrl: string;
  accessToken: string;
  documentId: number;
}

export interface ListSegmentsRequest {
  baseUrl: string;
  documentId: number;
}

export interface ListDocumentTasksRequest {
  baseUrl: string;
  accessToken: string;
  documentId: number;
}

export async function importDocument(
  request: ImportDocumentRequest,
): Promise<KnowledgeDocument> {
  const body = new FormData();
  body.append("file", request.file);
  body.append("knowledgeBaseId", request.knowledgeBaseId);
  if (request.docTitle?.trim()) {
    body.append("docTitle", request.docTitle.trim());
  }
  if (request.description?.trim()) {
    body.append("description", request.description.trim());
  }

  const response = await fetch(`${normalizeBaseUrl(request.baseUrl)}/documents/import`, {
    method: "POST",
    headers: {
      Authorization: `Bearer ${request.accessToken}`,
    },
    body,
  });

  if (!response.ok) {
    throw new Error(await readErrorMessage(response, "文档导入失败"));
  }

  return (await response.json()) as KnowledgeDocument;
}

export async function listDocuments(
  request: ListDocumentsRequest,
): Promise<KnowledgeDocument[]> {
  const params = new URLSearchParams({
    knowledgeBaseId: request.knowledgeBaseId,
  });
  const response = await fetch(
    `${normalizeBaseUrl(request.baseUrl)}/documents?${params.toString()}`,
    {
      method: "GET",
      headers: {
        Authorization: `Bearer ${request.accessToken}`,
      },
    },
  );

  if (!response.ok) {
    throw new Error(await readErrorMessage(response, "文档列表加载失败"));
  }

  return (await response.json()) as KnowledgeDocument[];
}

export async function splitDocument(
  request: SplitDocumentRequest,
): Promise<DocumentSplitResult> {
  const response = await fetch(
    `${normalizeBaseUrl(request.baseUrl)}/documents/${request.documentId}/split`,
    {
      method: "POST",
      headers: {
        "Content-Type": "application/json",
        Authorization: `Bearer ${request.accessToken}`,
      },
      body: JSON.stringify({
        splitType: "TITLE",
        chunkSize: 800,
        overlap: 80,
        titleLevel: 1,
      }),
    },
  );

  if (!response.ok) {
    throw new Error(await readErrorMessage(response, "文档切分失败"));
  }

  return (await response.json()) as DocumentSplitResult;
}

export async function listDocumentSegments(
  request: ListSegmentsRequest,
): Promise<KnowledgeSegment[]> {
  const response = await fetch(
    `${normalizeBaseUrl(request.baseUrl)}/documents/${request.documentId}/segments`,
    {
      method: "GET",
    },
  );

  if (!response.ok) {
    throw new Error(await readErrorMessage(response, "文档切分结果加载失败"));
  }

  return (await response.json()) as KnowledgeSegment[];
}

export async function listDocumentTasks(
  request: ListDocumentTasksRequest,
): Promise<DocumentTask[]> {
  const response = await fetch(
    `${normalizeBaseUrl(request.baseUrl)}/documents/${request.documentId}/tasks`,
    {
      method: "GET",
      headers: {
        Authorization: `Bearer ${request.accessToken}`,
      },
    },
  );

  if (!response.ok) {
    throw new Error(await readErrorMessage(response, "文档任务状态加载失败"));
  }

  return (await response.json()) as DocumentTask[];
}

async function readErrorMessage(
  response: Response,
  fallback: string,
): Promise<string> {
  const body = await response.text();
  if (!body) {
    return fallback;
  }

  try {
    const data = JSON.parse(body) as { detail?: string };
    return data.detail || fallback;
  } catch {
    return body;
  }
}

function normalizeBaseUrl(baseUrl: string): string {
  return baseUrl.replace(/\/+$/, "");
}
